File: src/main/kotlin/school/charset/app/domain/exercise/generator/Latin1Generator.kt
package school.charset.app.domain.exercise.generator

import school.charset.app.domain.encoding.CodePoint
import school.charset.app.domain.encoding.Codec
import school.charset.app.domain.encoding.Encoding
import school.charset.app.domain.exercise.Exercise
import school.charset.app.domain.exercise.ExerciseGenerationException
import school.charset.app.domain.exercise.Step

class Latin1Generator(
    private val codec: Codec,
    private val codePointGenerator: CodePointGenerator,
    private val byteArrayGenerator: ByteArrayGenerator,
) : EncodingExerciseGenerator {
    override val encoding = Encoding.Latin1

    override fun generateEncode(level: Int): Exercise.Encode {
        val latin1Level = parseLevel(level)
        val codePoint = codePointGenerator.randomLatin1(latin1Level)
        val steps = buildEncodeStepsFor(codePoint)
        return Exercise.Encode(codePoint, Encoding.Latin1, level, steps)
    }

    override fun generateDecode(level: Int): Exercise.Decode {
        val latin1Level = parseLevel(level)
        val bytes = byteArrayGenerator.randomLatin1(latin1Level)
        val codePoint = codec.decode(bytes, Encoding.Latin1)
        val steps = buildDecodeStepsFor(bytes)
        return Exercise.Decode(bytes, codePoint, Encoding.Latin1, level, steps)
    }

    fun buildEncodeStepsFor(codePoint: CodePoint): List<Step> {
        val bytes = codec.encode(codePoint, Encoding.Latin1)
        val byte = bytes[0].toInt() and 0xFF
        val binary = byte.toString(2).padStart(8, '0')
        return listOf(
            Step.Binary(expected = binary, length = 8),
            Step.HexBytes(expected = listOf(byte)),
        )
    }

    fun buildDecodeStepsFor(bytes: ByteArray): List<Step> {
        require(bytes.size == 1) {
            "Latin-1 decode requires exactly 1 byte, got ${bytes.size}"
        }
        val byte = bytes[0].toInt() and 0xFF
        val binary = byte.toString(2).padStart(8, '0')
        return listOf(
            Step.Binary(expected = binary, length = 8),
            Step.CodePointEntry(expected = byte),
        )
    }

    private fun parseLevel(level: Int): Latin1Level = Latin1Level.fromNumber(level)
        ?: throw ExerciseGenerationException(
            encoding = Encoding.Latin1,
            level = level,
            reason = "level must be one of: ${Latin1Level.validNumbers}",
        )
}

File: lib/charset/exercise/generator/latin1_generator.ex
defmodule Charset.Exercise.Generator.Latin1Generator do
  @moduledoc """
  Latin-1 exercises: single byte, identity between byte value and code point.

  `build_encode_steps_for/1` and `build_decode_steps_for/1` are also used by
  the sandbox (no level involved).
  """

  alias Charset.Encoding.Codec
  alias Charset.Exercise
  alias Charset.Exercise.Generator.CodePointGenerator
  alias Charset.Exercise.Generator.Levels
  alias Charset.Exercise.GenerationError
  alias Charset.Exercise.Step

  @spec generate_encode!(pos_integer()) :: Exercise.t()
  def generate_encode!(level) do
    tier = parse_level!(level)
    code_point = CodePointGenerator.random_latin1(tier)
    Exercise.encode(code_point, :latin1, level, build_encode_steps_for(code_point))
  end

  @spec generate_decode!(pos_integer()) :: Exercise.t()
  def generate_decode!(level) do
    tier = parse_level!(level)
    bytes = Codec.encode!(CodePointGenerator.random_latin1(tier), :latin1)
    code_point = Codec.decode!(bytes, :latin1)
    Exercise.decode(bytes, code_point, :latin1, level, build_decode_steps_for(bytes))
  end

  @spec build_encode_steps_for(integer()) :: [Step.t()]
  def build_encode_steps_for(code_point) do
    <<byte>> = Codec.encode!(code_point, :latin1)

    [
      Step.Binary.new!(binary_string(byte, 8), 8),
      Step.HexBytes.new!([byte])
    ]
  end

  @spec build_decode_steps_for(binary()) :: [Step.t()]
  def build_decode_steps_for(<<byte>>) do
    [
      Step.Binary.new!(binary_string(byte, 8), 8),
      Step.CodePointEntry.new!(byte)
    ]
  end

  defp binary_string(value, length) do
    value |> Integer.to_string(2) |> String.pad_leading(length, "0")
  end

  defp parse_level!(level) do
    case Levels.parse(:latin1, level) do
      {:ok, tier} ->
        tier

      :error ->
        raise GenerationError,
          encoding: :latin1,
          level: level,
          reason: "level must be one of: #{Levels.valid_numbers(:latin1)}"
    end
  end
end
